source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -94,40 +92,59 @@ void Filter::onComplete(Filters::Common::ExtAuthz::CheckStatus status) {
-                                             cluster_->statsScope(),
-                                             EMPTY_STRING,
-                                             enumToInt(Http::Code::Forbidden),
-                                             true,
-                                             EMPTY_STRING,
-                                             EMPTY_STRING,
-                                             EMPTY_STRING,
-                                             EMPTY_STRING,
-                                             false};
+    Http::CodeUtility::ResponseStatInfo info{
+        config_->scope(), cluster_->statsScope(), EMPTY_STRING, response->status_code, true,
+        EMPTY_STRING,     EMPTY_STRING,           EMPTY_STRING, EMPTY_STRING,          false};
     Http::CodeUtility::chargeResponseStat(info);
     break;
   }
 
   // We fail open/fail close based of filter config
   // if there is an error contacting the service.
-  if (status == CheckStatus::Denied ||
-      (status == CheckStatus::Error && !config_->failureModeAllow())) {
-    Http::HeaderMapPtr response_headers{new Http::HeaderMapImpl(*getDeniedHeader())};
-    callbacks_->encodeHeaders(std::move(response_headers), true);
+  if (response->status == CheckStatus::Denied ||
+      (response->status == CheckStatus::Error && !config_->failureModeAllow())) {
+
+    if (response->body && response->body->length()) {
+      callbacks_->encodeHeaders(getHeaderMap(response), false);
+      callbacks_->encodeData(*response->body.get(), true);
+    } else {
+      callbacks_->encodeHeaders(getHeaderMap(response), true);
+    }
+
     callbacks_->requestInfo().setResponseFlag(
         RequestInfo::ResponseFlag::UnauthorizedExternalService);
   } else {
-    if (config_->failureModeAllow() && status == CheckStatus::Error) {
+    if (config_->failureModeAllow() && response->status == CheckStatus::Error) {
       // Status is Error and yet we are allowing the request. Click a counter.
       cluster_->statsScope().counter("ext_authz.failure_mode_allowed").inc();
     }
+
     // We can get completion inline, so only call continue if that isn't happening.
     if (!initiating_call_) {
+      if (response->status == CheckStatus::OK) {
+        for (const auto& header : response->headers) {
+          request_headers_->setReferenceKey(header.first, header.second);
+        }
+      }
       callbacks_->continueDecoding();
     }
   }
 }
 
+Http::HeaderMapPtr Filter::getHeaderMap(const Filters::Common::ExtAuthz::ResponsePtr& reponse) {
+  Http::HeaderMapPtr header_map;
+  if (reponse->status_code != enumToInt(Http::Code::Forbidden)) {
+    header_map = std::make_unique<Http::HeaderMapImpl>(
+        Http::HeaderMapImpl{{Http::Headers::get().Status, std::to_string(reponse->status_code)}});
+  } else {
+    header_map = std::make_unique<Http::HeaderMapImpl>(*getDeniedHeader());
+  }
+  for (const auto& header : reponse->headers) {
```

Comment:
I wonder if it make sense to have a utility or method on header map to handle these copies, this pattern comes up a fair bit.

Reply:
👍